### PR TITLE
refactor: implement window creation

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

**Comment by marijnvdwerf** [comment]
Most of it should be alright/reviewable, I just need some help with inserting and `memmove`ing windows.

**Comment by AaronVanGeffen** [comment]
Do we still need the pad_all struct/union around the window structure? It appears MSVC really doesn't like the combination of default initialisers and unions ([error C2926](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-2/compiler-errors-c2900-through-c3499?view=vs-2019)).

File: src/openloco/ui/WindowManager.cpp [hunk 1]
```diff
@@ -541,36 +579,294 @@ namespace openloco::ui::WindowManager
...
+        window.type = type;
+        window.flags = flags;
+        if ((flags & window_flags::flag_12) != 0)
+        {
+            window.flags |= window_flags::white_border_mask;
+            audio::play_sound(audio::sound_id::open_window, origin.x + size.width / 2);
+        }
+        else if (((flags & window_flags::stick_to_back) == 0) && ((flags & window_flags::stick_to_front) == 0) && ((flags & window_flags::flag_13) == 0))
+        {
+            // FIXME: Same as previous case, but need a nice way to make the if clear
```
Comment (AaronVanGeffen) [inline src/openloco/ui/WindowManager.cpp]: Perhaps you could introduce temporary boolean variables, e.g. `has_flag_12`, `does_stick_to_back` to simplify the if-expression.

File: src/openloco/window.h [hunk 2]
```diff
@@ -320,51 +323,58 @@ namespace openloco::ui
             uint8_t pad_all[0x88E];
```
Comment (AaronVanGeffen) [inline src/openloco/window.h]: I think we should just get rid of this line and the encapsulating union entirely.

File: src/openloco/ui/WindowManager.cpp [hunk 3]
```diff
@@ -541,36 +579,294 @@ namespace openloco::ui::WindowManager
...
+                continue;
+
+            return false;
+        }
+
+        return true;
+    }
+
+    // 0x004C9F27
+    static window* foundSpace(
```
Comment (AaronVanGeffen) [inline src/openloco/ui/WindowManager.cpp]: `foundSpace` sounds a bit vague to me. I realise it's temporary until we can use a std container, but I think `emptyWindowSlot` or something might be clearer?

File: src/openloco/ui/WindowManager.cpp (hunk 1 above)
Comment (AaronVanGeffen) [reply src/openloco/ui/WindowManager.cpp]: The booleans have been introduced, but the two cases have yet to be merged.

File: src/openloco/ui/WindowManager.cpp [hunk 4]
```diff
@@ -541,36 +586,292 @@ namespace openloco::ui::WindowManager
...
+            || (!stickToBack && !stickToFront && !hasFlag13))
+        {
+            window.flags |= window_flags::white_border_mask;
+            audio::play_sound(audio::sound_id::open_window, origin.x + size.width / 2);
+        }
+
+        window.event_handlers = events;
+
+        size_t length = _windowsEnd - (_windows + dstIndex);
+        memmove(_windows + dstIndex + 1, _windows + dstIndex, length * sizeof(ui::window));
```
Comment (marijnvdwerf) [inline src/openloco/ui/WindowManager.cpp]: length can be 0, is this an issue?

Comment (AaronVanGeffen) [reply src/openloco/ui/WindowManager.cpp]: No, it should just be a no-op if the size argument is 0.

**Pull request merged**